infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/strategy/ServersMapper.java

```diff
@@ -0,0 +1,181 @@
+                } else {
+                    refs.addAll(port2refs.get(binding.numPort));
+                }
+            }
+
+            if (refs == null) {
+                mapped.put(binding.rawPort, new ServerImpl(makeUrl(binding, null, null)));
+            } else {
+                for (String ref : refs) {
+                    ServerConfig cfg = configs.get(ref);
+                    mapped.put(ref, new ServerImpl(makeUrl(binding, cfg.getProtocol(), cfg.getPath())));
+                }
+            }
+        }
+        return mapped;
+    }
+
+    /**
+     * Makes url from given binding, protocol and path.
+     *
+     * <p>Examples:
+     * <pre>
+     * |------------------------------------------------------------------------|
+     * | binding           | protocol | path    | url                           |
+     * |------------------------------------------------------------------------|
+     * | 4011/tcp -> 32011 | http     | /api    | http://hostname:32011/api     |
+     * | 4012/tcp -> 32012 | wss      | connect | wss://hostname:32012/connect  |
+     * | 4013     -> 32013 | https    |         | https://hostname:32013        |
+     * | 4014/upd -> 32014 |          |         | upd://hostname:32014          |
+     * | 4015     -> 32015 |          |         | tcp://hostname:32015          |
+     * |------------------------------------------------------------------------|
+     * </pre>
+     */
+    private String makeUrl(Binding binding, String protocol, String path) {
+        if (protocol == null) {
+            if (binding.protocol == null) {
+                protocol = "tcp";
+            } else {
+                protocol = binding.protocol;
+            }
+        }
+
+        // null -> "", "path" -> "/path"
+        if (path == null) {
+            path = "";
+        } else if (!path.startsWith("/")) {
+            path = '/' + path;
+        }
+
+        return protocol + "://" + hostname + ':' + binding.hostPort + path;
+    }
+
+    /** Retrieves numPort -> single binding map. */
+    private static List<Binding> getBindings(ContainerInfo container) {
+        NetworkSettings settings = container.getNetworkSettings();
+        if (settings == null)
+            return Collections.emptyList();
+
+        Map<String, List<PortBinding>> ports = settings.getPorts();
+        if (ports == null)
+            return Collections.emptyList();
+
+        return ports.entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue().size() == 1)
+                    .map(entry -> new Binding(entry.getKey(), entry.getValue().get(0)))
+                    .collect(Collectors.toList());
+    }
+
+    /** Structured representation of docker port -> single port binding mapping. */
+    private static class Binding {
+
+        /** The rawPort numPort in form 4011/tcp or 4011 if not protocol is not defined. */
+        final String rawPort;
```

Comment:
I think docker API always returns ports in format `4411/tcp`. Can you elaborate on this?